File: Gain Model/web/implosion_timeline.js
// Time-resolved ICF implosion timeline — JavaScript port of implosion_timeline.py.
//
// Same contract as the Python reference: simulate(design) -> animation-ready time
// series (t, R, T, rhoR, gain_t) + scalar outcomes + IGNITES/MARGINAL/QUENCHED
// verdict. Runs entirely in the browser (no backend) to drive the interactive
// dashboard. Physics: Bosch-Hale burn ODE, the coupled-gain pipeline, the multimode
// RT mix penalty, and a tuned gas-cushion trajectory — see the .py for derivations.
//
// Verified against the Python reference by web/verify.mjs.

import { DATA } from "./timeline_data.js";

// ---- physical constants (SI) ----
const KB = 1.380649e-23;
const KEV_J = 1.602176634e-16;
const E_ALPHA = 3.5e6 * 1.602176634e-19;
const MBAR = ((2.014 + 3.016) / 2) * 1.66053907e-27; // mean DT ion mass
const GAMMA = 5.0 / 3.0;
const C_BREM = 5.34e-37;
const RHO_R_HALF = 0.30;

// ---- gain_model constants ----
const E_FUSION = 17.6e6 * 1.602176634e-19;
const M_DT_PAIR = (2.014 + 3.016) * 1.66053907e-27;
const ETA_HOHLRAUM = 0.14, ETA_ROCKET = 0.075;
const RHOR_REF = 1.20, CR_REF = 35.0, ALPHA_REF = 2.0, RHO_HS = 300.0, N_TAU_BURN = 0.42;

// ---- coupled_gain (drive pressure + hydro-fit hot-spot T) ----
const drivePressure = (E_J) => DATA.P_DRIVE_NIF * Math.pow(E_J / DATA.E_LASER_NIF, DATA.DRIVE_EXP);
const hydroThs = (P) => Math.max(DATA.HYDRO_SLOPE * P + DATA.HYDRO_INT, 0.05);

// ---- rt_mix constants ----
const SIGMA_SURF_REF = 2.1e-8, SPEC_SLOPE = 1.0, ELL_CUT = 35.0;
const ACCEL_ADIA_EXP = 2.0, K_IFAR = 9.7, P_MIX = 2.5, V_ABL_REF = 5.0e3;
// rt_mechanics (Takabe ablative growth)
const G_ACCEL = 1.9e14, ATWOOD = 0.9, ALPHA_T = 0.9, BETA_T = 3.0, T_ACCEL = 1.85e-9;

const V_IMP_NIF = 3.7e5;
const R0 = 1.0e-3;

// ---- small numeric helpers ----
const linspace = (a, b, n) => Array.from({ length: n }, (_, i) => a + (b - a) * (i / (n - 1)));
function interp(x, xp, fp) {
  // numpy.interp: linear, clamped to endpoints. xp ascending.
  if (x <= xp[0]) return fp[0];
  const n = xp.length;
  if (x >= xp[n - 1]) return fp[n - 1];
  let lo = 0, hi = n - 1;
  while (hi - lo > 1) { const m = (lo + hi) >> 1; if (xp[m] <= x) lo = m; else hi = m; }
  const t = (x - xp[lo]) / (xp[hi] - xp[lo]);
  return fp[lo] + t * (fp[hi] - fp[lo]);
}
const argmin = (a) => { let k = 0; for (let i = 1; i < a.length; i++) if (a[i] < a[k]) k = i; return k; };
const amax = (a) => a.reduce((m, v) => (v > m ? v : m), -Infinity);

// ---- DT fusion reactivity <sigma v> (Bosch & Hale 1992), input keV -> m^3/s ----
function reactivityDT(T_keV) {
  const T = Math.min(Math.max(T_keV, 0.2), 100.0);
  const BG = 34.3827, MRC2 = 1.124656e6;
  const C1 = 1.17302e-9, C2 = 1.51361e-2, C3 = 7.51886e-2;
  const C4 = 4.60643e-3, C5 = 1.35e-2, C6 = -1.0675e-4, C7 = 1.366e-5;
  const theta = T / (1.0 - (T * (C2 + T * (C4 + T * C6))) / (1.0 + T * (C3 + T * (C5 + T * C7))));
  const xi = Math.pow((BG * BG) / (4.0 * theta), 1.0 / 3.0);
  const sv_cm3 = C1 * theta * Math.sqrt(xi / (MRC2 * T * T * T)) * Math.exp(-3.0 * xi);
  return sv_cm3 * 1e-6;
}
const fAlpha = (rhoR) => rhoR / (rhoR + RHO_R_HALF);
const pAlpha = (n_i, T, rhoR) => fAlpha(rhoR) * 0.25 * n_i * n_i * reactivityDT(T) * E_ALPHA;
const pBrem = (n_i, T) => C_BREM * n_i * n_i * Math.sqrt(T);
const soundSpeed = (T_keV) => Math.sqrt((GAMMA * 2.0 * (T_keV * KEV_J)) / MBAR);

// ---- 0-D hot-spot burn: integrate [T, n_i] over one confinement time (RK4) ----
function burn(rho, rhoR, T0_keV, n_tau = 1.0) {
  const rho_si = rho * 1000.0;
  const n_i0 = rho_si / MBAR;
  const Rhs = (rhoR * 10.0) / rho_si;
  const tau_c = (n_tau * Rhs) / (4.0 * soundSpeed(T0_keV));

  const deriv = (T, n_i) => {
    T = Math.max(T, 0.2); n_i = Math.max(n_i, 0.0);
    const heat = pAlpha(n_i, T, rhoR) - pBrem(n_i, T);
    const rate = 0.25 * n_i * n_i * reactivityDT(T);
    const dn = -2.0 * rate;
    const dT = n_i > 0 ? (heat - 3.0 * (T * KEV_J) * dn) / (3.0 * n_i * KB) / (KEV_J / KB) : 0.0;
    return [dT, dn];
  };

  const nOut = 400, sub = 20, steps = nOut * sub, dt = tau_c / steps;
  const t = new Array(nOut + 1), Ts = new Array(nOut + 1), ni = new Array(nOut + 1);
  let T = T0_keV, n = n_i0;
  t[0] = 0; Ts[0] = T; ni[0] = n;
  let out = 1;
  for (let i = 1; i <= steps; i++) {
    const [k1T, k1n] = deriv(T, n);
    const [k2T, k2n] = deriv(T + 0.5 * dt * k1T, n + 0.5 * dt * k1n);
    const [k3T, k3n] = deriv(T + 0.5 * dt * k2T, n + 0.5 * dt * k2n);
    const [k4T, k4n] = deriv(T + dt * k3T, n + dt * k3n);
    T += (dt / 6.0) * (k1T + 2 * k2T + 2 * k3T + k4T);
    n += (dt / 6.0) * (k1n + 2 * k2n + 2 * k3n + k4n);
    T = Math.max(T, 0.2); n = Math.max(n, 0.0);
    if (i % sub === 0) { t[out] = i * dt; Ts[out] = T; ni[out] = n; out++; }
  }
  return { t, T: Ts, n_i: ni, tau_c, burnup: 1.0 - n / n_i0 };
}

function evaluateGain(E_J, fuel_kg, CR, adiabat) {
  const T_hs = hydroThs(drivePressure(E_J));
  const rhoR = RHOR_REF * (CR / CR_REF) ** 2 * (ALPHA_REF / adiabat);
  const b = burn(RHO_HS, rhoR, T_hs, N_TAU_BURN);
  const n_react = (b.burnup * fuel_kg) / M_DT_PAIR;
  const Y = n_react * E_FUSION;
  return { gain: Y / E_J, T_hs, yield_J: Y, burnup: b.burnup };
}

// ---- multimode RT mix penalty ----
const ELL = Array.from({ length: DATA.RT_ELLMAX }, (_, i) => i + 1);
const SEED = (() => {
  const s = ELL.map((l) => Math.pow(l, -SPEC_SLOPE));
  const norm = Math.sqrt(s.reduce((a, v) => a + v * v, 0));
  return s.map((v) => v / norm);
})();
const CUTOFF = ELL.map((l) => Math.exp(-((l / ELL_CUT) ** 2)));

function gammaAblative(lam, Va) {
  const k = (2.0 * Math.PI) / lam;
  return Math.max(ALPHA_T * Math.sqrt(ATWOOD * G_ACCEL * k) - BETA_T * k * Va, 0.0);
}
function accelFeedthrough(adiabat) {
  const Va = V_ABL_REF * Math.pow(adiabat / ALPHA_REF, ACCEL_ADIA_EXP);
  return ELL.map((l) => {
    const lam = (2.0 * Math.PI * DATA.RT_R0) / l;
    const nE = Math.min(gammaAblative(lam, Va) * T_ACCEL, 6.0);
    return Math.exp(nE);
  });
}
function gfAt(cr) {
  const CRs = DATA.RT_CR, GF = DATA.RT_GF;
  const c = Math.min(Math.max(cr, CRs[0]), CRs[CRs.length - 1]);
  return ELL.map((_, j) => interp(c, CRs, GF.map((row) => row[j])));
}
const ifarAmp = (cr, ad) => Math.exp(K_IFAR * (Math.sqrt((cr / CR_REF) * (ALPHA_REF / ad)) - 1.0));

function mixState(cr, adiabat, sigma) {
  const accel = accelFeedthrough(adiabat), gf = gfAt(cr);
  const bp = cr / CR_REF, ifar = ifarAmp(cr, adiabat);
  let s2 = 0;
  for (let j = 0; j < ELL.length; j++) {
    const eta = sigma * SEED[j] * accel[j] * gf[j] * CUTOFF[j] * ifar * bp;
    s2 += eta * eta;
  }
  const sigma_mix = Math.sqrt(s2);
  return { sigma_mix, f_mix: sigma_mix / (DATA.RT_R0 / cr) };
}
const mixPenalty = (cr, ad, sigma = SIGMA_SURF_REF) =>
  Math.pow(Math.max(0.0, 1.0 - Math.min(mixState(cr, ad, sigma).f_mix, 1.0)), P_MIX);

const MIX_REF = mixPenalty(35.0, 2.0, SIGMA_SURF_REF);
const mixFactor = (d) => Math.min(mixPenalty(d.CR, d.adiabat, d.surf_nm * 1e-9) / MIX_REF, 1.0 / MIX_REF);

// ---- implosion trajectory R(t): coast + decelerate on a gas cushion tuned to CR ----
const implosionVelocity = (E_MJ) => V_IMP_NIF * Math.sqrt(E_MJ / 2.05);

function trajectory(v_imp, CR, n = 1400) {
  const R_min_target = R0 / CR, exp = 3.0 * GAMMA;
  const tmax = (2.6 * R0) / v_imp;
  const run = (K, steps) => {
    const dt = tmax / steps;
    const T = new Array(steps + 1), R = new Array(steps + 1);
    let r = R0, v = -v_imp, rmin = R0, broke = false;
    for (let i = 0; i <= steps; i++) {
      T[i] = i * dt; R[i] = r;
      const a = K * Math.pow(R0 / Math.max(r, 1e-9), exp);
      const rm = r + 0.5 * dt * v, vm = v + 0.5 * dt * a;
      const am = K * Math.pow(R0 / Math.max(rm, 1e-9), exp);
      r = r + dt * vm; v = v + dt * am;
      rmin = Math.min(rmin, r);
      if (r < R0 / 400.0) { for (let k = i; k <= steps; k++) { T[k] = k * dt; R[k] = r; } broke = true; break; }
    }
    return { T, R, rmin, broke };
  };
  let lo = 1e6, hi = 1e18;
  for (let it = 0; it < 60; it++) {
    const K = Math.sqrt(lo * hi);
    if (run(K, 500).rmin < R_min_target) lo = K; else hi = K;
  }
  const K = Math.sqrt(lo * hi);
  const { T, R } = run(K, n);
  return { t: T, R };
}

// ---- the simulation ----
export function simulate(d, npts = 600) {
  const E_J = d.E_MJ * 1e6;
  const v_imp = implosionVelocity(d.E_MJ);

  const T_stag = hydroThs(drivePressure(E_J));
  const rhoR_stag = RHOR_REF * (d.CR / CR_REF) ** 2 * (ALPHA_REF / d.adiabat);
  const rho_stag = RHO_HS;
  const mix_pen = mixPenalty(d.CR, d.adiabat, d.surf_nm * 1e-9);
  const rhoR_eff = rhoR_stag * Math.max(Math.pow(mix_pen, 0.3), 0.05);

  const gain = evaluateGain(E_J, d.fuel_ug * 1e-9, d.CR, d.adiabat).gain * mixFactor(d);
  const yield_J = gain * E_J;

  const tr = trajectory(v_imp, d.CR);
  const istag = argmin(tr.R);
  const t_stag = tr.t[istag], R_min = tr.R[istag];

  const b = burn(rho_stag, rhoR_eff, T_stag, N_TAU_BURN);
  const phi = b.n_i.map((v) => 1.0 - v / b.n_i[0]);
  const phiEnd = Math.max(phi[phi.length - 1], 1e-9);
  const ignites = amax(b.T) > 1.6 * T_stag;

  const t_end = tr.t[tr.t.length - 1] + b.tau_c;
  const t = linspace(0, t_end, npts);
  const R = t.map((tt) => interp(tt, tr.t, tr.R));
  const comp = R.map((r) => Math.min(Math.max(R_min / Math.max(r, 1e-12), 0), 1));
  const T = comp.map((c) => T_stag * c * c);
  const rho = comp.map((c) => rho_stag * c * c * c);
  const rhoR = comp.map((c) => rhoR_stag * c * c);
  const gain_t = new Array(npts).fill(0);
  const T_end_burn = Math.max(b.T[b.T.length - 1], T_stag);

  for (let i = 0; i < npts; i++) {
    if (t[i] >= t_stag && t[i] <= t_stag + b.tau_c) {
      const tb = t[i] - t_stag;
      T[i] = interp(tb, b.t, b.T);
      rho[i] = rho_stag; rhoR[i] = rhoR_stag;
      gain_t[i] = (gain * interp(tb, b.t, phi)) / phiEnd;
    } else if (t[i] > t_stag + b.tau_c) {
      gain_t[i] = gain;
      const c = Math.min(Math.max(R_min / Math.max(R[i], 1e-12), 0), 1);
      T[i] = T_end_burn * c * c;
    }
  }

  const verdict = mix_pen < 0.25 ? "QUENCHED" : ignites && gain >= 1.0 ? "IGNITES" : "MARGINAL";
  return {
    t, R, T, rho, rhoR, gain_t,
    t_stag, R_min, T_stag, T_peak: amax(T), rhoR_stag,
    mix_penalty: mix_pen, burnup: b.burnup, gain, yield_MJ: yield_J / 1e6,
    ignites, verdict, v_imp, tau_c: b.tau_c,
  };
}

export const PRESETS = {
  "NIF 2022 (first ignition, gain 1.5)": { E_MJ: 2.05, fuel_ug: 210.0, CR: 35.0, adiabat: 2.0, surf_nm: 21.0 },
  "NIF 2025 (record, gain 4.1)": { E_MJ: 2.08, fuel_ug: 225.0, CR: 44.0, adiabat: 1.7, surf_nm: 2.0 },
};

File: Gain Model/implosion_timeline.py
r"""
Time-resolved implosion timeline -- one animatable model of an ICF shot.

The repo's other models each give a *number* (a gain, a mix penalty) or a single
stage. This one stitches the stages into a single evolution in time -- the thing you
would actually watch:

    coast  ->  convergence  ->  stagnation  ->  ignition (or fizzle)  ->  disassembly

driven by the same physics the rest of the repo is built on and anchored to the same
NIF calibration, so the two preset shots reproduce their real gains:

    * implosion trajectory R(t)   -- shell coasts inward, decelerates on the hot-spot
      gas cushion, stagnates at R_min = R0/CR, and rebounds (a tuned 2-var ODE);
    * compression      T(t), rho(t), rho*R(t) rise adiabatically as the fuel converges,
      reaching the coupled-model stagnation state (hot-spot T from the 1-D hydro fit);
    * burn             the repo's 0-D hot-spot ODE (Bosch-Hale reactivity, alpha
      heating vs bremsstrahlung) seeded at stagnation -- it runs away (ignites) or
      decays (fizzles); its T(t) is the ignition spike;
    * mix              the multimode RT penalty thins the confining areal density, so
      an aggressive / rough-surface design fizzles where a clean one ignites.

This is the Python reference. A JavaScript port drives the interactive dashboard;
`simulate()` returns everything the animation needs (time series + verdict + gain).

Presets: NIF N221204 (Dec 2022, gain 1.54) and the NIF record shot (Apr 2025, gain
4.13) -- same ~2 MJ laser, a better capsule (more compression, less mix), exactly the
lesson the earlier steps draw.

Run:  python3 implosion_timeline.py
Deps: numpy, scipy  (matplotlib only for the figure)
"""

import os
import importlib.util
from dataclasses import dataclass

import numpy as np


def _load(mod, fname):
    spec = importlib.util.spec_from_file_location(mod, os.path.join(os.path.dirname(__file__), fname))
    m = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(m)
    return m


cg = _load("coupled_gain", "coupled_gain.py")
mix = _load("rt_mix", "rt_mix.py")
gm = cg.gm                      # gain_model (efficiencies, stagnation scalings, burn)
hs = gm.hotspot_0d              # 0-D burn physics (reactivity, p_alpha, p_brem, burn)

R0 = 1.0e-3                     # initial capsule / shell radius [m]
GAMMA = 5.0 / 3.0
V_IMP_NIF = 3.7e5              # implosion velocity at the NIF reference energy [m/s]


@dataclass
class Design:
    E_MJ: float        # laser energy on target [MJ]
    fuel_ug: float     # DT fuel mass [micrograms]  (NIF ~ 210 ug = 0.21e-6 kg)
    CR: float          # convergence ratio R0/R_min
    adiabat: float     # in-flight adiabat
    surf_nm: float     # capsule surface finish (RMS) [nm]  -- HIGH-mode RT seed (mix)
    drive_asym_pct: float = 0.0   # P2 drive (pressure-wave) asymmetry [%] -- LOW-mode RT


# preset shots -- capsule knobs chosen so the coupled physics reproduces the real gains
PRESETS = {
    "NIF 2022 (first ignition, gain 1.5)": Design(2.05, 210.0, 35.0, 2.00, 21.0),
    "NIF 2025 (record, gain 4.1)":         Design(2.08, 225.0, 44.0, 1.70, 2.0),
}

MIX_REF = mix.mix_penalty(35.0, 2.0, mix.SIGMA_SURF_REF)   # nominal mix penalty (~0.85)


def mix_factor(d):
    """Mix penalty relative to the NIF-like nominal (as in robust_design_ml)."""
    return min(mix.mix_penalty(d.CR, d.adiabat, d.surf_nm * 1e-9) / MIX_REF, 1.0 / MIX_REF)


# P2 drive (pressure-wave) asymmetry -> yield-over-clean. Low-mode drive asymmetry
# seeds an l=2 perturbation that spherical convergence + deceleration RT amplify into a
# peanut-shaped hot spot, degrading yield. Calibrated to Rayleigh-Taylor/hohlraum_asymmetry.py:
# 1% P2 -> YOC ~0.72, 2% -> ~0.35, 5% -> ~0.02 (the ignition cliff, not a fitted curve).
ASYM_A0, ASYM_N = 0.016, 2.3


def asym_yoc(drive_asym_pct):
    """Yield-over-clean for a given P2 drive asymmetry [%] (1.0 = perfectly symmetric)."""
    a2 = max(drive_asym_pct, 0.0) / 100.0
    return 1.0 / (1.0 + (a2 / ASYM_A0) ** ASYM_N)


def implosion_velocity(E_MJ):
    """Laser energy -> implosion velocity [m/s] (rocket coupling ~ sqrt(energy))."""
    return V_IMP_NIF * (E_MJ / 2.05) ** 0.5


def _trajectory(v_imp, CR, n=1400):
    """Shell coasts in at v_imp and decelerates on a gas cushion tuned to reach R_min=R0/CR.

    Returns (t, R): the radius history through stagnation and part of the rebound.
    R'' = K (R0/R)^(3 gamma) outward; K bisected so min(R) == R0/CR.
    """
    R_min_target = R0 / CR
    exp = 3.0 * GAMMA

    def run(K, tmax, steps):
        dt = tmax / steps
        r, v = R0, -v_imp
        T = np.empty(steps + 1); R = np.empty(steps + 1)
        rmin = R0
        for i in range(steps + 1):
            T[i], R[i] = i * dt, r
            a = K * (R0 / max(r, 1e-9)) ** exp
            # RK2 (midpoint) on (r, v); r' = v, v' = a(r)
            rm = r + 0.5 * dt * v
            vm = v + 0.5 * dt * a
            am = K * (R0 / max(rm, 1e-9)) ** exp
            r = r + dt * vm
            v = v + dt * am
            rmin = min(rmin, r)
            if r < R0 / 400.0:
                R[i:] = r; T[i:] = np.arange(i, steps + 1) * dt
                break
        return T, R, rmin

    # time to stagnation ~ R0 / v_imp; integrate a bit past for the rebound
    tmax = 2.6 * R0 / v_imp
    lo, hi = 1e6, 1e18
    for _ in range(60):                      # bisection on K to hit the target CR
        K = np.sqrt(lo * hi)
        _, _, rmin = run(K, tmax, 500)
        if rmin < R_min_target:              # too much convergence -> stiffen cushion
            lo = K
        else:
            hi = K
    K = np.sqrt(lo * hi)
    return run(K, tmax, n)[:2]


def simulate(d, npts=600):
    """Run one shot. Returns the full time series + scalar outcomes for the animation."""
    E_J = d.E_MJ * 1e6
    v_imp = implosion_velocity(d.E_MJ)

    # --- stagnation state, from the validated coupled model ---
    T_stag = cg.hydro_T_hs(cg.drive_pressure(E_J))                      # hot-spot T [keV]
    rhoR_stag = gm.RHOR_REF * (d.CR / gm.CR_REF) ** 2 * (gm.ALPHA_REF / d.adiabat)  # g/cm^2
    rho_stag = gm.RHO_HS                                                # g/cc
    mix_pen = mix.mix_penalty(d.CR, d.adiabat, d.surf_nm * 1e-9)        # yield multiplier
    asym = asym_yoc(getattr(d, "drive_asym_pct", 0.0))                  # P2 (low-mode) YOC
    # both instabilities thin the confining hot spot -> a smaller burning core
    rhoR_eff = rhoR_stag * max((mix_pen * asym) ** 0.3, 0.03)

    # --- scalar gain: the validated pipeline (coupled x mix) x drive-asymmetry YOC ---
    gain = cg.evaluate(cg.Design(E_J, d.fuel_ug * 1e-9, d.CR, d.adiabat))["gain"] * mix_factor(d) * asym
    yield_J = gain * E_J

    # --- implosion trajectory R(t) ---
    t_tr, R_tr = _trajectory(v_imp, d.CR)
    istag = int(np.argmin(R_tr))
    t_stag, R_min = t_tr[istag], R_tr[istag]

    # --- burn at stagnation (repo's 0-D hot-spot ODE gives the T(t) ignition/fizzle shape) ---
    b = gm.burn(rho_stag, rhoR_eff, T_stag, n_tau=gm.N_TAU_BURN)
    burnup = b["burnup"]
    tau_c = b["tau_c"]
    phi = 1.0 - b["n_i"] / b["n_i"][0]                                  # burned fraction vs burn-time
    ignites = b["T"].max() > 1.6 * T_stag                              # clear thermal runaway

    # --- assemble a single timeline: compression -> burn spike -> disassembly ---
    # non-uniform grid: dense through stagnation + the (tens-of-ps) burn so the
    # ignition spike is resolved; index-based playback then slows down there for free.
    t_end = t_tr[-1] + tau_c
    n1, n2 = int(npts * 0.47), int(npts * 0.36)
    n3 = npts - n1 - n2
    pre = np.linspace(0.0, t_stag, n1, endpoint=False)          # coast (0 .. <t_stag)
    mid = np.linspace(t_stag, t_stag + tau_c, n2)               # stagnation + burn (dense)
    post = np.linspace(t_stag + tau_c, t_end, n3 + 1)[1:]       # disassembly
    t = np.concatenate([pre, mid, post])
    R = np.interp(t, t_tr, R_tr)
    comp = np.clip(R_min / np.maximum(R, 1e-12), 0.0, 1.0)              # 0 (start) -> 1 (stagnation)
    T = T_stag * comp ** 2                                              # adiabatic compression heating
    rho = rho_stag * comp ** 3
    rhoR = rhoR_stag * comp ** 2
    gain_t = np.zeros_like(t)

    # overlay the burn window [t_stag, t_stag + tau_c]: real T(t) and ramping gain
    win = (t >= t_stag) & (t <= t_stag + tau_c)
    tb = t[win] - t_stag
    T[win] = np.interp(tb, b["t"], b["T"])
    rho[win] = rho_stag; rhoR[win] = rhoR_stag
    gain_t[win] = gain * np.interp(tb, b["t"], phi) / max(phi[-1], 1e-9)
    after = t > t_stag + tau_c
    gain_t[after] = gain
    # post-burn disassembly: expansion cooling as the shell rebounds
    T[after] = max(b["T"][-1], T_stag) * np.clip(R_min / np.maximum(R[after], 1e-12), 0.0, 1.0) ** 2

    clean = mix_pen * asym                                             # combined RT survival
    verdict = "QUENCHED" if clean < 0.25 else ("IGNITES" if ignites and gain >= 1.0 else "MARGINAL")
    return dict(
        t=t, R=R, T=T, rho=rho, rhoR=rhoR, gain_t=gain_t,
        t_stag=t_stag, R_min=R_min, T_stag=T_stag, T_peak=float(T.max()),
        rhoR_stag=rhoR_stag, mix_penalty=mix_pen, asym_yoc=asym,
        drive_asym_pct=float(getattr(d, "drive_asym_pct", 0.0)), burnup=burnup,
        gain=gain, yield_MJ=yield_J / 1e6, ignites=bool(ignites), verdict=verdict,
        v_imp=v_imp, tau_c=tau_c,
    )


def report():
    print("=" * 70)
    print("  IMPLOSION TIMELINE  --  coast -> stagnation -> ignition/fizzle")
    print("=" * 70)
    for name, d in PRESETS.items():
        s = simulate(d)
        print(f"\n  {name}")
        print(f"    inputs : E={d.E_MJ} MJ  fuel={d.fuel_ug} ug  CR={d.CR}  "
              f"adiabat={d.adiabat}  surface={d.surf_nm} nm")
        print(f"    v_imp {s['v_imp']/1e3:.0f} km/s  ->  stagnation at "
              f"{s['t_stag']*1e9:.2f} ns,  R_min {s['R_min']*1e6:.1f} um")
        print(f"    hot-spot T {s['T_stag']:.1f} keV   peak T {s['T_peak']:.1f} keV   "
              f"rho*R {s['rhoR_stag']:.2f} g/cm^2   mix {s['mix_penalty']*100:.0f}%")
        print(f"    burn-up {s['burnup']*100:.1f}%   yield {s['yield_MJ']:.2f} MJ   "
              f"GAIN {s['gain']:.2f}   [{s['verdict']}]")
    print("=" * 70)


def make_figure(fname="implosion_timeline.png"):
    try:
        import matplotlib.pyplot as plt
    except Exception as e:
        print(f"[figure skipped: {e}]")
        return
    fig, ax = plt.subplots(1, 3, figsize=(15, 4.6))
    colors = {"NIF 2022 (first ignition, gain 1.5)": "tab:blue",
              "NIF 2025 (record, gain 4.1)": "tab:red"}
    sims = {name: simulate(d) for name, d in PRESETS.items()}

    for name, s in sims.items():
        c = colors[name]; lab = name.split(" (")[0]
        ax[0].plot(s["t"] * 1e9, s["R"] * 1e6, c, lw=2, label=lab)
        ax[1].plot(s["t"] * 1e9, s["T"], c, lw=2, label=lab)
        ax[2].plot(s["t"] * 1e9, s["gain_t"], c, lw=2, label=f"{lab}  (gain {s['gain']:.1f})")

    ax[0].set_xlabel("time [ns]"); ax[0].set_ylabel("shell radius [µm]")
    ax[0].set_title("implosion trajectory\ncoast → stagnation → rebound"); ax[0].legend(fontsize=8)
    ax[1].axhline(4.3, color="gray", ls=":", lw=1); ax[1].text(0.1, 4.6, "ignition ~4.3 keV", fontsize=8)
    ax[1].set_xlabel("time [ns]"); ax[1].set_ylabel("hot-spot T [keV]")
    ax[1].set_title("hot-spot temperature\ncompression + ignition spike"); ax[1].legend(fontsize=8)
    ax[2].axhline(1.0, color="k", ls=":", lw=1); ax[2].text(0.1, 1.1, "breakeven", fontsize=8)
    ax[2].set_xlabel("time [ns]"); ax[2].set_ylabel("cumulative gain")
    ax[2].set_title("gain accumulates during the burn"); ax[2].legend(fontsize=8)

    fig.suptitle("ICF implosion timeline: two NIF shots, same laser, different capsule", fontsize=13)
    fig.tight_layout(); fig.savefig(fname, dpi=130)
    print(f"Saved figure -> {fname}")


def main():
    report()
    make_figure()


if __name__ == "__main__":
    main()

# ----------------------------------------------------------------------------
# NOTES ----------------------------------------------------------------------
# * Reduced and stitched, not a rad-hydro solve: a tuned gas-cushion trajectory, an
#   adiabatic compression law for T/rho/rho*R, and the repo's 0-D burn spliced in at
#   stagnation -- all anchored to the coupled-gain calibration so the presets land on
#   the real NIF gains. The point is a faithful *shape* to animate, not a design code.
# * The mix penalty enters as a reduction of the confining areal density, so a rough
#   or over-aggressive capsule fizzles in the same burn ODE that ignites a clean one.
# * simulate() is the contract for the JS port: identical inputs -> identical arrays.
# ----------------------------------------------------------------------------
